### ISSUE TITLE
Allow configurable trust domain in Istio identities

### ISSUE THREAD

**Issue by elevran** [istio/istio#8661]
**Describe the feature request**
Allow use of configurable domain in Istio identities.

**Describe alternatives you've considered**
currently all identities are hard coded to use the `cluster.local` domain.

**Additional context**
#7849 has initial work in Citadel, but needs further work in other components:

- [ ] add option support in Helm charts (incl istio-remote)
- [ ] changes in istioctl to kube-inject `verify_subject_alt_name` (e.g., [here](https://github.com/istio/istio/blob/master/pkg/bootstrap/bootstrap_config.go#L46))
- [ ] change [hard coded places](https://github.com/istio/istio/search?p=1&q=spiffe%3A%2F%2Fcluster.local&unscoped_q=spiffe%3A%2F%2Fcluster.local), except tests
- [ ] collect identity related items into a new package (e.g., `security/pkg/identity`) to avoid duplication

**Comment by loewenstein** [comment]
@elevran Not sure how the below link fits to the description.

> * changes in istioctl to kube-inject `verify_subject_alt_name` (e.g., [here](https://github.com/istio/istio/blob/master/pkg/bootstrap/bootstrap_config.go#L46))

As a general note, we should probably use perm links to code as much as possible.

**Comment by loewenstein** [comment]
@elevran I think we should broaden the search for hardcoded values to 'cluster.local'. [Sourcegraph](https://about.sourcegraph.com/) has support for more flexible queries like [this query](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/istio/istio%24+cluster.local+-file:_test%5C.go%24+-file:%5E%5C.circleci/.*%24+-file:%5Ebin/.*%24+-file:testdata/.*%24)

Another thought I had was actually not excluding tests from 'cluster.local' removal (although probably handling separate and with lower priority). The benefit would be that a simple `grep -rnw -e 'cluster.local'` would show that we are done. Not completely sure though if it is really worth the effort. Thoughts?

**Comment by elevran** [comment]
@loewenstein the above were copied from the comments to previous PR (#7849), for reference. We can certainly change as needed. 

> @elevran Not sure how the below link fits to the description.
> 
> > * changes in istioctl to kube-inject `verify_subject_alt_name` (e.g., [here](https://github.com/istio/istio/blob/master/pkg/bootstrap/bootstrap_config.go#L46))
> 
> As a general note, we should probably use perm links to code as much as possible.

agree. It was meant to refer to:
```go
   defaultPilotSan = []string{
		"spiffe://cluster.local/ns/istio-system/sa/istio-pilot-service-account"}
```

The reasoning behind not changing the tests is to confirm that we are backward compatible. We could also add tests to confirm the use of non-default domains.

**Comment by loewenstein** [comment]
cc @holgero

**Comment by c0d1ngm0nk3y** [comment]
see #8976 for addressing those issues

**Comment by myidpt** [comment]
@elevran Please see my comment in https://github.com/istio/istio/pull/9059#pullrequestreview-163458139.
For your first item, I think we should store the identityDomain in ConfigMap and have each deployment refer to it as env variables. The benefits are:
1. You have a centralized place to track the identityDomain of the cluster. It also avoids mismatching of the identityDomain arg, especially for manually deployed components without using helm.
2. You don't need to reconfigure every single deployment when the idenityDomain changes during migration. Especially when the identity domain is automatically managed by a control plane that does not necessary have the helm templates.

**Comment by loewenstein** [comment]
@elevran @myidpt should this be milestone 1.1? As is, master has code and configuration that does not work. At the very least https://github.com/istio/istio/pull/8800 should be merged before a 1.1 release.

Also, we should agree on the config name (identityDomain vs. trustDomain; the latter being what SPIFFE uses in the spec)

**Comment by myidpt** [comment]
@loewenstein I doubt if we can make this feature valuable in 1.1.
For https://github.com/istio/istio/pull/8800, can someone help to update the PR?

I think we should use trustDomain instead. We also use trustDomain in our discussion.

**Comment by loewenstein** [comment]
I know planned release is approaching rapidly (@hklai when is release-1.1 branch going to be cut approximately?), but isn't most of the work already done in https://github.com/istio/istio/pull/7849? At least I had the feeling this issue is mostly about left-overs (@elevran)?
Things like moving config from helm values to config map are not strictly required, are they? Backwards compatibility could be achieved by taking the value from helm values.yaml and putting it into the config map.

Not sure about 
```
changes in istioctl to kube-inject verify_subject_alt_name (e.g., here)
```
the other checklist items do have PRs in flight or in review. Anything missing in the checklist?

**Comment by hklai** [comment]
See
https://docs.google.com/document/d/1fkCaNYTPudSBUka3BBitzNNQA2Uzeq0vFe7E7_n2WUc/edit

We will create a burn down list to figure out where we are, and the plan is
to cut release-1.1 by end of the week.

**Comment by c0d1ngm0nk3y** [comment]
I am a bit confused. Please correct me if I am wrong, but #8800 does actually nothing. So we agree that `trustDomain` is a better term? Then we should at very least rename it. Once the wording `identityDomain` is out and established, it could get confusing.

I agree that the ConfigMap should be another PR. My suggestion:

1. Renaming
2. Fix feature (not only for citadel)
3. Use of ConfigMap

What do you think?

**Comment by myidpt** [comment]
Thank you folks for this effort. I think this implementation is done and we have verified it's working. Can someone please add a corresponding task to https://preliminary.istio.io/docs/tasks/security/? I can do it as well if no one takes this work :)
@elevran @axel7born @loewenstein
